ufl/core/external_operator.py

```diff
@@ -0,0 +1,134 @@
+from ufl.coefficient import Coefficient
+from ufl.core.ufl_type import ufl_type
+from ufl.constantvalue import as_ufl
+from ufl.log import error
+from ufl.finiteelement.finiteelement import FiniteElement
+from ufl.finiteelement.mixedelement import VectorElement, TensorElement
+from ufl.functionspace import FunctionSpace
+from ufl.domain import default_domain
+
+
+@ufl_type(inherit_indices_from_operand=0, is_terminal=True, is_differential=True)
+class ExternalOperator(Coefficient):
+
+    # Slots are disabled here because they cause trouble in PyDOLFIN
+    # multiple inheritance pattern:
+    _ufl_noslots_ = True
+
+    def __init__(self, *operands, function_space, derivatives=None, count=None):
+        r"""
+        :param operands: operands on which acts the :class:`ExternalOperator`.
+        :param function_space: the :class:`.FunctionSpace`,
+        or :class:`.MixedFunctionSpace` on which to build this :class:`Function`.
+        Alternatively, another :class:`Function` may be passed here and its function space
+        will be used to build this :class:`Function`.  In this case, the function values are copied.
```

Comment:
Should I just say `Coefficient` or remove this bit? Because picking up the function space from a `firedrake.Function` (inheriting from Coefficient) is something done in Firedrake and I don't know if other entities using UFL also have this small mechanism.